kion/resource_label_test.go

```diff
@@ -0,0 +1,162 @@
+package kion
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	hc "github.com/kionsoftware/terraform-provider-kion/kion/internal/kionclient"
+)
+
+const (
+	resourceTypeLabel = "kion_label"
+	resourceNameLabel = "label1"
+)
+
+var (
+	initialTestLabel = hc.LabelCreate{
+		Color: "#123abc",
+		Key:   "environment",
+		Value: "production",
+	}
+	updatedTestLabel = hc.LabelCreate{
+		Color: "#abcdef",
+		Key:   "env",
+		Value: "prod",
+	}
+)
+
+func TestAccResourceLabelCreate(t *testing.T) {
+	config := testAccLabelGenerateResourceDeclaration(&initialTestLabel)
+	hc.PrintHCLConfig(config) // Print the generated HCL configuration
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccLabelCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check:  resource.ComposeTestCheckFunc(testAccLabelCheckResource(&initialTestLabel)...),
+			},
+		},
+	})
+}
```

Comment:
I'm struggling to understand what this is actually testing.  I'm new to terraform tests, but this seems to just be testing the `testAccLabelGenerateResourceDeclaration` and `testAccLabelCheckResource` methods.  What am I missing?

Reply:
I'm not sure you're missing anything.  I borrowed inspiration from the OU test and then looked at what AWS is doing. Essentially, they test whether the resource can be created, if it can be modified, and deleted.

1. `TestAccResourceLabelCreate:`
This checks that the `kion_label` resource can be successfully created by generating the HCL and applying it, while `testAccLabelCheckResource` makes sure the values are what's expected.

2. `TestAccResourceLabelUpdate`:
This checks that the `kion_label` resource can be updated by first applying the initial configuration and updating it, and then checking to see if it was updated.

3.  `TestAccResourceLabelDelete`:
This checks that the `kion_label` resource can be deleted by first applying the initial configuration and then emptying the HCL to "delete it."

With all this being said, I guess just generating HCL without checking for the specific kv "Works" it makes sense to add more specific tests based on the proposed HCL config.

Reply:
How are you running tests?  Do you have `KION_URL` set to a local instance?  Any particular setup needed, or do you just run against a blank install?

Reply:
This checks for the HCL configuration.  VSCode has tests built in, I just click a test function, and it runs.

<img width="740" alt="Screenshot 2024-06-07 at 7 09 17 PM" src="https://github.com/kionsoftware/terraform-provider-kion/assets/112982230/eab233a4-cbff-4d6b-a66b-1e551c58808c">

The output of hitting that test button is:

```
Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestAccResourceLabelCreate$ github.com/kionsoftware/terraform-provider-kion/kion

ok  	github.com/kionsoftware/terraform-provider-kion/kion	0.423s
```

Reply:
The terraform SDK test stuff is very non-standard go tests, so VSCode isn't going to run it correctly.  Here's some good documentation: https://developer.hashicorp.com/terraform/plugin/sdkv2/testing/acceptance-tests#running-acceptance-tests

Essentially you must run `go test -v` with the `-v` modifier, per [this](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/resource/testing.go#L690).  And you must set `TF_ACC=true` per [this](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/resource/testing.go#L686-L688).

Interestingly, despite what the docs say, I didn't see tests _fail_ without these two things.  I saw what looked likes tests running, but the test code inside `TestStep` doesn't actually run.